### DISCUSSION TITLE
Compressed quaternion decoding

### DISCUSSION THREAD

**Discussion by AkashJohnSubash** [bitcraze/discussions#553] [Q&A] [answered]
I'm using an off-board controller to generate an optimal trajectory, which defines the state vector as : 
[Positions (m), attitude (quaternion), linear velocities (m/s), angular velocities (rad/sec)]

To get this observed onboard state from the Crazyflie, I'm synchronously logging the `StateEstimateZ `group, with a single callback registered at 100ms.

- does the cflib support decoding this compressed quaternion format. i.e CompressedQaut -> [qw, qx, qy, qz] ?
If not, I'd apprecite instructions on how :) 
I could take the original [qw, qx, qy, qz] values from `StateEstimate `group, but I'd like to fetch all of it from a single module in firmware so that they are in sync, and be able to log it using a single callback

- does the period of the callbacks depend on the size of the data packet ?
 i.e I'd like to reduce the callback period as much as possible to have the most accurate state estimate

**Comment by krichardsson** [comment, marked as answer]
> does the cflib support decoding this compressed quaternion format. i.e CompressedQaut -> [qw, qx, qy, qz] ?
If not, I'd apprecite instructions on how :)

Not that I know of, but it could be useful. I think the compressed logging was implemented by [Crazyswam](https://crazyswarm.readthedocs.io/en/latest/) and I don't think it has been used that much from the python lib? You will have to re-implement the [decoding function](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/interface/quatcompress.h#L65-L84) in python, there is some bit fiddeling but it should be possible. It is possible that the [`struct.unpack`](https://docs.python.org/3/library/struct.html) feature in python can be useful, not sure.

> do the period of the callbacks depend on the size of the data packet ?
i.e I'd like to reduce the callback period as much as possible to have the most accurate state estimate

No, as long as there is bandwidth available in the link you should get the rate you configure. One log "event" is always transmitted in one radio packet, so the size not that important. This is also why there is a limit on how many log variables that can be configured into one log block.
